source/server-selection/server-selection.rst

```diff
@@ -1180,11 +1185,85 @@ server without an explicit ``start_request`` (or comparable) method call.
 Outside a legacy "request" API, drivers MUST use server selection for each
 individual read operation.
 
-Reference Implementation
-========================
-
-The single-threaded reference implementation is the Perl master branch (work
-towards v1.0.0).  The multi-threaded reference implementation is TBD.
+Logging
+-------
+Please refer to the `logging specification <../logging/logging.rst>`_ for
+details on logging implementations in general, including log levels, log
+components, and structured versus unstructured logging.
+
+Drivers MUST support logging of server selection information via the
+following log messages. These messages MUST use the ``serverSelection`` log
+component.
+
+The types used in the structured message definitions below are demonstrative,
+and drivers MAY use similar types instead so long as the information is present
+(e.g. a double instead of an integer, or a string instead of an integer if the
+structured logging framework does not support numeric types.)
+
+"Waiting for suitable server to become available" message
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+This message MUST be logged at ``info`` level. It MUST be emitted on the occasions
+specified either in `Multi-threaded or asynchronous server selection`_ or
+`Single-threaded server selection`_, depending on which algorithm the driver
+implements.
+
+In order to avoid generating redundant log messages where all information besides
+the remaining time is identical, this message MUST not be repeatedly emitted for an
+operation unless the topology description changes.
+
+This message MUST contain the following key-value pairs:
+
+.. list-table::
+   :header-rows: 1
+   :widths: 1 1 1
+
+   * - Key
+     - Suggested Type
+     - Value
+
+   * - message
+     - String
+     - "Waiting for suitable server to become available"
+
+   * - selector
+     - String
+     - String representation of the selector being used to select the server. This can be
+       a read preference or an application-provided custom selector. The exact content of
+       is flexible depending on what the driver is able to log. At minimum, when the selector
+       is a read preference this string MUST contain all components of the read preference,
+       and when it is an application-provided custom selector the string MUST somehow indicate
+       that it is a custom selector.
+
+   * - requestId
+     - Int
+     - The driver-generated ID for the request a server is being selected for. Optional; this
+       field will not be present in cases where server selection does not correspond to an operation,
+       e.g. when the driver performs server selection to check if a `deployment supports sessions
+       <../sessions/driver-sessions.rst#how-to-check-whether-a-deployment-supports-sessions>`__.
+
+   * - operation
```

Comment:
Currently in .Net operation information is not propagated into server selection. Do you think we'll need it for the rest of server selection logs?

Reply:
what I currently have proposed locally is that all the additional messages (started, succeeded, failed) would have the operation name and operation ID as well.